node/setup.go

```diff
@@ -42,16 +42,34 @@ import (
 
 const readHeaderTimeout = 10 * time.Second
 
-// GenesisDocProvider returns a GenesisDoc.
+// ChecksummedGenesisDoc combines a GenesisDoc together with its
+// SHA256 checksum.
+type ChecksummedGenesisDoc struct {
+	GenesisDoc     *types.GenesisDoc
+	Sha256Checksum []byte
+}
+
+// GenesisDocProvider returns a GenesisDoc together with its SHA256 checksum.
 // It allows the GenesisDoc to be pulled from sources other than the
 // filesystem, for instance from a distributed key-value store cluster.
-type GenesisDocProvider func() (*types.GenesisDoc, error)
+type GenesisDocProvider func() (ChecksummedGenesisDoc, error)
 
 // DefaultGenesisDocProviderFunc returns a GenesisDocProvider that loads
 // the GenesisDoc from the config.GenesisFile() on the filesystem.
 func DefaultGenesisDocProviderFunc(config *cfg.Config) GenesisDocProvider {
-	return func() (*types.GenesisDoc, error) {
```

Comment:
I wonder whether instead of changing the return value of this function (which is used in many other places), we can not move the checksum logic into the `LoadStateFromDBOrGenesisProvider` to be executed only once if there is no hash in the DB or if we have a flag provided via cli? Basically move it to line 576  and put the logic of reading the file and computing the sum there?

Reply:
> I wonder whether instead of changing the return value of this function (which is used in many other places), we can not move the checksum logic into the `LoadStateFromDBOrGenesisProvider` to be executed only once if there is no hash in the DB or if we have a flag provided via cli? Basically move it to line 576 and put the logic of reading the file and computing the sum there?

I have not come across any other places where the `GenesisDocProvider` signature was used. My concern is about eventually reworking this for streamability: the provider should not need to store the whole file content in RAM to parse the JSON or calculate the checksum. This signature allows implementing it with an incremental parser (e.g. `json.Decoder` from the standard library) and an incremental SHA-256 hasher. I have left a FIXME comment about this in the default provider function. Maybe I should turn it into a backlog issue.

Reply:
Filed away as #1302

Reply:
`DefaultGenesisProvider` is being used a lot although mostly to create a new node so technically it should be fine. But I am not sure whether it is used by users as well .

So instead of changing a public interface for a future optimization, could we not do one of two things:
- Leave this until we do the actual optimization and then we can even create a `DefaultStreamingDocProvider` or
- Keep the `genesisProvider` returning just the doc and have a special `createChecksum` function that could afterwards be altered to not read the file but rather stream through it based on a given path. This function is then called to get the hash (the fact that the genesisProvider has it in memory does not change anything as we anyways still keep it in memory).

Reply:
Also, we cannot really do this without changing the signature and contract for `GenesisDocProvider` in some way: in main, it only returns the parsed doc, and both retrieval or JSON content from the source and parsing is supposed to happen inside the provider function.

Reply:
There is one backward-compatible alternative: add a checksum field to the `GenesisDoc` struct and mandate every genesis doc provider set it with the computed checksum. But: 1) it feels conceptually wrong; 2) this special field has to be fenced away from normal handling of the `GenesisDoc` fields by the JSON marshal and unmarshal implementations.

Reply:
Yes, you are right it might be wrong.. What abut the solution of changing the signature of `LoadStateFromDBOrGenesisDocProvider` by adding it a path to the genesis file (as we will anyways change this function signature to pass the hash from command line)? Both are breaking, just not sure which is more (as in breaking the way people are used to use these things).

Reply:
The SDK uses `GenesisDocProvider` in three places (if I didn't miss any): [here](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/server/start.go#L398), [here](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/server/start.go#L362), and [here](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/server/cmt_cmds.go#L404).

If we include a changelog explaining the signature change, I doesn't seem highly disruptive. Thoughts?

Reply:
> If we include a changelog explaining the signature change, I doesn't seem highly disruptive. Thoughts?

I support this. But I do need to add a changelog, after all.